images/tensorflow-notebook/cuda12/Dockerfile

```diff
@@ -0,0 +1,34 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+ARG REGISTRY=quay.io
+ARG OWNER=jupyter
+ARG BASE_CONTAINER=$REGISTRY/$OWNER/scipy-notebook
+FROM $BASE_CONTAINER
+
+LABEL maintainer="Jupyter Project <jupyter@googlegroups.com>"
+
+# Fix: https://github.com/hadolint/hadolint/wiki/DL4006
+# Fix: https://github.com/koalaman/shellcheck/wiki/SC3014
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# Install CUDA libs and cuDNN with mamba
+RUN mamba install --yes -c nvidia/label/cuda-12.3.2 \
```

Comment:
Is there a chance not to hardcode minor.patch version here?
Also, 12.4 was released.

Reply:
I used 12.3, because it is listed in the [tested build configurations](https://www.tensorflow.org/install/source#gpu). There seems to be no label like `cuda-12`.  Without label, i. e. `-c nvidia` I got the latest CUDA version, which is 12.4, currently. It worked, but I thought it is risky. A new CUDA release might be incompatible (I guess not before 13.x) and it seems we do not have a unit test e. g. to check the output of `python -c "import tensorflow as tf; print(tf.config.list_physical_devices('GPU'))"`. Should we add a unit test like this?
Secondly, the [pytorch-notebook](https://github.com/jupyter/docker-stacks/blob/07dc2d08734b384ab7370956d47909bda9bc928b/images/pytorch-notebook/cuda12/Dockerfile#L16) fixes the minor version as well.

Reply:
> I used 12.3, because it is listed in the tested build configurations.

There is a chance that the table might be slightly outdated.

> There seems to be no label like cuda-12. Without label, i. e. -c nvidia I got the latest CUDA version, which is 12.4, currently.

Can we use something like `nvidia/label/cuda-12.*`?

> Should we add a unit test like this?

I am not sure if this test will work with a regular GitHub-hosted ubuntu runner.
Also, we currently don't have a way to run a test for `variant` image (but it's not difficult to add something like this).

Reply:
> There is a chance that the table might be slightly outdated.

I looked at the table just a few hours after the release and it was up to date. But it won't ever be tested against a newer version.

> Can we use something like nvidia/label/cuda-12.*?

No, there is no such label. [Here](https://anaconda.org/nvidia/cuda-nvcc/labels) is a list.

> I am not sure if this test will work with a regular GitHub-hosted ubuntu runner.

Right, that is always problematic. Sorry.

Reply:
> > Can we use something like nvidia/label/cuda-12.*?
> 
> No, there is no such label. [Here](https://anaconda.org/nvidia/cuda-nvcc/labels) is a list.

I meant maybe mamba supports label regex (I have no idea if it does or not).
In that case we will be fine with existing tags and won’t need to hardcode some particular version.

Reply:
Despite using labels, I noticed all package versions are also in the main label. The version in the [list of labels](https://anaconda.org/nvidia/cuda-nvcc/labels) is a bit strange. So the labels are formatted in major.minor.patch and the version in major.minor.build. Usually, if patch increments, the build number just continues:

- label: cuda-11.6.0, version: 11.6.55
- label: cuda-11.6.1, version: 11.6.112
- label: cuda-11.6.2, version: 11.6.124

There is one exception: labels cuda-11.4.3 and cuda-11.4.4 both have version 11.4.152.
Nevertheless, I just tried:

- `mamba install -c nvidia 'cuda-nvcc<13'` and got version 12.4.99 (which is also in label cuda-12.4.0)
- `mamba install -c nvidia 'cuda-nvcc<12'` and got version 11.8.89 (which is also in label cuda-11.8.0)
- `mamba install -c nvidia 'cuda-nvcc<11.5'` and got version 11.4.152 (which is also in labels cuda-11.4.3 and cuda-11.4.4)
- `mamba install -c nvidia 'cuda-nvcc=12.3'` and got version 12.3.107 (which is also in label cuda-12.3.2)

So, we could use `'cuda-nvcc<13'`, to reduce maintenance work (avoid updating the version for every TensorFlow release), but these are not officially tested by TensorFlow (not sure, if there can occur incompatibilities with new minor versions). Using something like `'cuda-nvcc=12.3'` is more work (still avoiding the patch version), but officially [tested](https://www.tensorflow.org/install/source#gpu).

Reply:
I'm ok with `cuda-nvcc=12.3`, please add the comment why we choose this version though.

Reply:
This did not work, since the dependencies have no limitation in their versions. So using `cuda-nvcc=12.3` with the nvidia channel resulted in a mixture of 12.3 and 12.4. NVIDIA is quite sloppy in their packaging.

Then I noticed, that [cudnn from the nvidia channel](https://anaconda.org/nvidia/cudnn) is outdated. Apparently they dropped the support 3 years ago. The [cudnn from conda-forge](https://anaconda.org/conda-forge/cudnn) is quite up-to-date, but there is no CUDA 12 build yet, only CUDA 11.8.

So, I would like to go with the new installation method [supported by TensorFlow](https://www.tensorflow.org/install/pip), which is basically just `pip install tensorflow[and-cuda]`. This also has the advantage, that the installed CUDA version is always the [officially tested](https://www.tensorflow.org/install/source#gpu) version – so less maintenance for you. Usually the path to the nvidia libs should be found automatically, but in 2.16.1 there seems to be a [bug](https://github.com/tensorflow/tensorflow/issues/63362), so we have to add them. I prepared an activation script for that. Also `LD_LIBRARY_PATH` is not polluted with this method, because the paths from the pip installation contain only the nvidia libs. Before we added `${CONDA_DIR}/lib/`, which contains quite a lot libraries.

Reply:
Ok, let’s try it